cmd/activator/main.go

```diff
@@ -133,6 +150,70 @@ func main() {
 		logger.Fatalw("Failed to create stats reporter", zap.Error(err))
 	}
 
+	// Set up signals so we handle the first shutdown signal gracefully.
+	stopCh := signals.SetupSignalHandler()
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, defaultResyncInterval)
+	servingInformerFactory := servinginformers.NewSharedInformerFactory(servingClient, defaultResyncInterval)
+	endpointInformer := kubeInformerFactory.Core().V1().Endpoints()
+	revisionInformer := servingInformerFactory.Serving().V1alpha1().Revisions()
+
+	go endpointInformer.Informer().Run(stopCh)
+	go revisionInformer.Informer().Run(stopCh)
+
+	params := queue.BreakerParams{QueueDepth: breakerQueueDepth, MaxConcurrency: breakerMaxConcurrency, InitialCapacity: 0}
+
+	// Return the number of endpoints, 0 if no endpoints are found.
+	endpointsGetter := func(revID activator.RevisionID) (int32, error) {
+		endpoints, err := endpointInformer.Lister().Endpoints(revID.Namespace).Get(revID.Name)
+		if errors.IsNotFound(err) {
+			return 0, nil
+		}
+		if err != nil {
+			return 0, err
+		}
+		addresses := activator.EndpointsAddressCount(endpoints.Subsets)
+		return int32(addresses), nil
+	}
+
+	// Return the revision from the observer.
+	revisionGetter := func(revID activator.RevisionID) (*v1alpha12.Revision, error) {
+		rev, err := revisionInformer.Lister().Revisions(revID.Namespace).Get(revID.Name)
+		if err != nil {
+			return nil, err
+		}
+		return rev, err
+	}
+
+	throttlerParams := activator.ThrottlerParams{BreakerParams: params, Logger: logger, GetEndpoints: endpointsGetter, GetRevision: revisionGetter}
+	throttler := activator.NewThrottler(throttlerParams)
```

Comment:
So the throttler takes into consideration the concurrency of the target revision, but I wonder if/how it takes into consideration that the activator itself may be horizontally scaled?

If the activator deployment has `N` pods, then we should be throttling the traffic we send to each Pod of the Revision to `{concurrency} / N`, vs. just `{concurrency}`.

Maybe you are doing this and I just haven't gotten there yet, but I wanted to make sure I wrote it down.

Reply:
At the moment we don't do it. Would be happy to send a separate PR to enable this. Right now we do have only a single activator instance (there are > 1 in the tests through), I think it would be enough for the scope of initial throttling enablement.

Reply:
I'm fine with separating it, but wanted to confirm this is a problem (and also have it factor into your thinking as you chase this).  Can you open an issue to track and add it to the project dashboard (or DM me a link and I can?).

thanks.

Reply:
Here is the issue: https://github.com/knative/serving/issues/3165